pkgs/development/compilers/dmd/default.nix

```diff
@@ -64,10 +64,13 @@ stdenv.mkDerivation rec {
       patchShebangs .
   '';
 
-  postPatch = stdenv.lib.optionalString stdenv.hostPlatform.isLinux ''
+  postPatch = ''
+    # Fix runnable/extra-files/cpp11.cpp:2:10: fatal error: 'cstddef' file not found
+    substituteInPlace dmd/test/runnable/cpp11.d --replace " -std=c++11" " -std=c++11 -x c++"
```

Comment:
I don't get why "-x c++" needs to be added to fix the error. What does this parameter even do I haven't found any documentation for that on https://clang.llvm.org/docs/ClangCommandLineReference.html .
Is it setting the language to c++? Why isn't it enough to use std=c++11?

Reply:
Because `CC=clang` and to get clang to handle C++ you can use `-x` to override the language.

Reply:
@Geod24 You worked on those tests lately, does it make sense to fix that upstream?

Reply:
This is an upstream bug. Apple clang will automatically switch to C++, just like gcc. But, we use the vanilla clang, which works as a C compiler when called as clang and C++ when called as clang++.

Reply:
Ah there was even a CC=$CXX in the previous default.nix, this probably fixes the issue too.

@lionello  The cpp11 test files have been moved upstream already so this substitute fix would only work until the next version bump.
Better to fix it via clang++ instead.

Reply:
> @Geod24 You worked on those tests lately, does it make sense to fix that upstream?

Definitely. I had this issue with druntime. Invoking `clang` and `clang++` behaves differently.
But we should be using `HOST_CXX` and `CXX` flavor everywhere, e.g. see: https://github.com/dlang/dmd/blob/f2cdcd5fea5d1418155a9008d5d5cc2141a28eaf/src/posix.mak#L66-L71

Reply:
Right, when I put back the `CC=$CXX`, this test fails:
```
Running dshell tests
 ... dshell/test6952.d
==============================
Test 'dshell/test6952.d' failed. The logged output:
[COMPILE_TEST] '/tmp/d/dmd/generated/linux/release/64/dmd' '-conf=' '-m64' '-fPIC' '-odtest_results/dshell/test6952' '-oftest_results/dshell/test6952/run' '-I=/tmp/d/dmd/test/dshell' '-I=test_results/dshell/test6952' '-I=/tmp/d/dmd/test/tools/dshell_prebuilt' '-i' '-i=-dshell_prebuilt' 'test_results/dshell_prebuilt.o' '/tmp/d/dmd/test/dshell/test6952.d'
[RUN_TEST] 'test_results/dshell/test6952/run'
core.exception.AssertError@/tmp/d/dmd/test/dshell/test6952.d(32): Couldn't find 'cc' in compiler output:
predefs   DigitalMars Posix linux ELFv1 CRuntime_Glibc CppRuntime_Gcc LittleEndian D_Version2 all D_SIMD D_InlineAsm_X86_64 X86_64 D_LP64 D_PIC assert D_ModuleInfo D_Exceptions D_TypeInfo D_HardFloat
binary    /tmp/d/dmd/generated/linux/release/64/dmd
version   v2.091.0

config
DFLAGS    (none)
parse     main
importall main
import    object	(dshell/extra-files/test6952/object.d)
semantic  main
semantic2 main
semantic3 main
code      main
function  main._start
function  main.main
g++ test_results/dshell/test6952/main.o -o test_results/dshell/test6952/main -g -m64 -nostartfiles -nostdlib -nodefaultlibs -lpthread -lm -lrt -ld
```
because that test expects to find either `cc` or `gcc` in the verbose output.

Reply:
I'm not sure I follow. The original `CC=$CXX` seems wrong, it should be `CC=gcc` or `CC=clang`, although that bit is likely to trigger the test failure.
Contact me directly on dlang's Slack if you want to troubleshoot it together.

Reply:
To me it seems this linker driver comparison is very narrow.
clang, clang++, g++, ... are all valid drivers for this no?

It also seems that it's only usage is to find the correct line which later gets parsed.
Wouldn't it be possible to turn the result parsing around?
To get the line search for "-nostartfiles -nostdlib -nodefaultlibs" and then make sure that all other stuff is not in that line.

This would make the test shorter and more platform neutral.

Reply:
I already had to fix this test when packaging for Musl: https://github.com/dlang/dmd/commit/f5cf08e3af5dbb6679f4c7d826e1b0c1d00420fd

But I think the *proper* test would be to write a wrapper script that checks what arguments are passed to it [and pass that to DMD via `CC`](https://github.com/dlang/dmd/blob/c6ab9c400f54704a1287facc2de03d7a10c10e33/src/dmd/link.d#L461).

@lionello do you feel like an upstream PR ? If you do it I'll review/merge it ;)

Reply:
@Geod24 Can you please elaborate what this test checks?
I don't know what the function behind the test even provides. It seems to be in the C++ interop functionality, but what exactly?

@lionello I would say to bring this PR to the finish line you can just remove this test from the testsuite for now under linux with a comment about the error and the information that it might be fixed upstream in a future version.

Reply:
Not sure if I follow.
Seems that `dmd` uses the `CC` env var and that's what gets printed, so shouldn't that test simply look for that same value?

Reply:
Yes that would be another solution for me too.
I also don't understand why there needs to be a wrapper of something.

Reply:
https://github.com/dlang/dmd/pull/11053

Reply:
Thanks @Geod24.

@lionello I think you could apply that patch directly with applyPatches/fetchpatch for [1] without waiting for the next release.

[1] https://github.com/dlang/dmd/commit/14547ae37a97678a80a32fea4c76626bbd8595ca.patch

Reply:
Done